File: app/csharp/AccountingSystem.Application/Services/JournalService.cs
using AccountingSystem.Application.Exceptions;
using AccountingSystem.Domain.Entities;
using AccountingSystem.Application.Ports.In;
using AccountingSystem.Application.Ports.Out;

namespace AccountingSystem.Application.Services;

/// <summary>
/// 仕訳サービス実装
/// </summary>
public class JournalService : IJournalService
{
    private readonly IJournalRepository _journalRepository;

    public JournalService(IJournalRepository journalRepository)
    {
        _journalRepository = journalRepository;
    }

    public async Task<Journal> GetJournalByNoAsync(string journalNo)
    {
        var journal = await _journalRepository.FindByJournalNoAsync(journalNo);

        if (journal == null)
        {
            throw new JournalNotFoundException($"仕訳伝票番号 {journalNo} が見つかりません");
        }

        return journal;
    }

    public async Task<Journal> CreateJournalAsync(Journal journal)
    {
        ValidateJournal(journal);

        // 既存チェック
        var existing = await _journalRepository.FindByJournalNoAsync(journal.JournalNo);
        if (existing != null)
        {
            throw new DuplicateJournalException($"仕訳伝票番号 {journal.JournalNo} は既に存在します");
        }

        await _journalRepository.InsertAsync(journal);

        return (await _journalRepository.FindByJournalNoAsync(journal.JournalNo))!;
    }

    public async Task DeleteJournalAsync(string journalNo)
    {
        var existing = await _journalRepository.FindByJournalNoAsync(journalNo);
        if (existing == null)
        {
            throw new JournalNotFoundException($"仕訳伝票番号 {journalNo} が見つかりません");
        }

        await _journalRepository.DeleteByJournalNoAsync(journalNo);
    }

    public async Task<(decimal DebitTotal, decimal CreditTotal, bool IsBalanced)> ValidateBalanceAsync(string journalNo)
    {
        var existing = await _journalRepository.FindByJournalNoAsync(journalNo);
        if (existing == null)
        {
            throw new JournalNotFoundException($"仕訳伝票番号 {journalNo} が見つかりません");
        }

        var (debitTotal, creditTotal) = await _journalRepository.GetBalanceAsync(journalNo);
        var isBalanced = debitTotal == creditTotal;

        return (debitTotal, creditTotal, isBalanced);
    }

    public async Task<IReadOnlyList<Journal>> GetJournalsByFiscalYearAsync(int fiscalYear)
    {
        return await _journalRepository.FindByFiscalYearAsync(fiscalYear);
    }

    private static void ValidateJournal(Journal journal)
    {
        if (string.IsNullOrWhiteSpace(journal.JournalNo))
        {
            throw new InvalidJournalEntryException("仕訳伝票番号は必須です");
        }

        if (journal.Details == null || journal.Details.Count == 0)
        {
            throw new InvalidJournalEntryException("仕訳明細は1件以上必要です");
        }

        // 複式簿記の検証: 借方・貸方の合計が一致するか
        decimal debitTotal = 0;
        decimal creditTotal = 0;

        foreach (var detail in journal.Details)
        {
            if (detail.Items == null || detail.Items.Count == 0)
            {
                throw new InvalidJournalEntryException($"行番号 {detail.LineNumber} の貸借明細は1件以上必要です");
            }

            foreach (var item in detail.Items)
            {
                if (string.IsNullOrWhiteSpace(item.AccountCode))
                {
                    throw new InvalidJournalEntryException($"行番号 {detail.LineNumber} の勘定科目コードは必須です");
                }

                if (item.Amount < 0)
                {
                    throw new InvalidJournalEntryException($"行番号 {detail.LineNumber} の金額は0以上である必要があります");
                }

                if (item.DebitCreditFlag == "D")
                {
                    debitTotal += item.Amount;
                }
                else if (item.DebitCreditFlag == "C")
                {
                    creditTotal += item.Amount;
                }
                else
                {
                    throw new InvalidJournalEntryException($"行番号 {detail.LineNumber} の貸借区分は 'D'（借方）または 'C'（貸方）である必要があります");
                }
            }
        }

        if (debitTotal != creditTotal)
        {
            throw new InvalidJournalEntryException(
                $"借方合計（{debitTotal:N0}）と貸方合計（{creditTotal:N0}）が一致しません。複式簿記の原則に違反しています。");
        }
    }
}

File: app/fsharp/AccountingSystem.Application/Services/JournalService.fs
namespace AccountingSystem.Application.Services

open System
open System.Threading.Tasks
open AccountingSystem.Application.Exceptions
open AccountingSystem.Application.Port.In
open AccountingSystem.Application.Port.Out
open AccountingSystem.Domain.Models.Journal
open AccountingSystem.Domain.Types

/// <summary>
/// 仕訳サービス実装
/// ビジネスロジックとトランザクション管理
/// </summary>
type JournalService(journalRepository: IJournalRepository) =

    /// <summary>
    /// 貸借一致の検証
    /// </summary>
    let validateDoubleEntry (journal: Journal) : Result<unit, string> =
        let totalDebit = Journal.sumDebit journal |> Money.toDecimal
        let totalCredit = Journal.sumCredit journal |> Money.toDecimal

        if totalDebit <> totalCredit then
            Error $"貸借が一致していません。借方合計: {totalDebit}, 貸方合計: {totalCredit}"
        else
            Ok ()

    /// <summary>
    /// 仕訳の検証
    /// </summary>
    let validateJournal (journal: Journal) : Result<unit, string> =
        // 1. 明細が存在するか
        if journal.Lines.IsEmpty then
            Error "仕訳明細が必要です"
        // 2. 各明細に項目が存在するか
        elif journal.Lines |> List.exists (fun line -> line.Items.IsEmpty) then
            Error "各仕訳明細には少なくとも1つの項目が必要です"
        else
            // 3. 貸借一致の検証
            validateDoubleEntry journal

    interface IJournalUseCase with

        member _.GetJournalByVoucherNumberAsync(voucherNumber: int) : Task<Journal> =
            task {
                let! journal = journalRepository.FindByVoucherNumberAsync(string voucherNumber)

                return
                    journal
                    |> Option.defaultWith (fun () ->
                        raise (JournalNotFoundException(voucherNumber)))
            }

        member _.GetJournalsByDateRangeAsync(fromDate: DateTime, toDate: DateTime) : Task<Journal list> =
            task {
                if fromDate > toDate then
                    raise (ArgumentException("開始日は終了日より前である必要があります"))

                return! journalRepository.FindByDateRangeAsync(fromDate, toDate)
            }

        member _.CreateJournalAsync(journal: Journal) : Task<Journal> =
            task {
                // ビジネスルール検証
                match validateJournal journal with
                | Error msg -> raise (InvalidJournalEntryException(msg))
                | Ok () -> ()

                // 重複チェック
                let! existing = journalRepository.FindByVoucherNumberAsync(string journal.VoucherNumber.Number)

                match existing with
                | Some _ ->
                    return raise (BusinessRuleViolationException($"伝票番号 {journal.VoucherNumber.Number} は既に存在します"))
                | None ->
                    let! _ = journalRepository.SaveAsync(journal)
                    let! created = journalRepository.FindByVoucherNumberAsync(string journal.VoucherNumber.Number)
                    return
                        created
                        |> Option.defaultWith (fun () ->
                            raise (JournalNotFoundException(journal.VoucherNumber.Number)))
            }

        member _.UpdateJournalAsync(voucherNumber: int) (journal: Journal) : Task<Journal> =
            task {
                // 存在チェック
                let! existing = journalRepository.FindByVoucherNumberAsync(string voucherNumber)

                match existing with
                | None ->
                    return raise (JournalNotFoundException(voucherNumber))
                | Some _ ->
                    // ビジネスルール検証
                    match validateJournal journal with
                    | Error msg -> raise (InvalidJournalEntryException(msg))
                    | Ok () -> ()

                    // 伝票番号を保持して更新
                    let updatedJournal = { journal with VoucherNumber = VoucherNumber.Create(string voucherNumber) }
                    let! _ = journalRepository.UpdateAsync(updatedJournal)
                    let! updated = journalRepository.FindByVoucherNumberAsync(string voucherNumber)
                    return
                        updated
                        |> Option.defaultWith (fun () ->
                            raise (JournalNotFoundException(voucherNumber)))
            }

        member _.DeleteJournalAsync(voucherNumber: int) : Task<unit> =
            task {
                let! existing = journalRepository.FindByVoucherNumberAsync(string voucherNumber)

                match existing with
                | None ->
                    raise (JournalNotFoundException(voucherNumber))
                | Some _ ->
                    let! _ = journalRepository.DeleteAsync(string voucherNumber)
                    return ()
            }
